src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(T item, Node<T> prev, Node<T> next) {
+            this.item = item;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            addFirst(value);
+        } else {
+            addLast(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0) {
+            addFirst(value);
+        } else if (index == size) {
+            addLast(value);
+        } else {
+            checkIndex(index);
+            addMiddle(getNodeByIndex(index), value);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
+    }
+
+    private void addFirst(T value) {
+        if (size == 0) {
+            first = new Node<>(value, null, last);
+            return;
+        }
+        Node<T> newNode = new Node<>(value, null, first);
+        first.prev = newNode;
+        first = newNode;
+    }
+
+    private void addLast(T value) {
+        Node<T> newNode = new Node<>(value, null, null);
+        if (last == null) {
+            first.next = newNode;
+            newNode.prev = first;
+        } else {
+            last.next = newNode;
+            newNode.prev = last;
+        }
+        last = newNode;
+    }
+
+    private void addMiddle(Node<T> nodeBefore, T value) {
+        Node<T> newNode = new Node<>(value,nodeBefore.prev, nodeBefore);
+        nodeBefore.prev.next = newNode;
+        nodeBefore.prev = newNode;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToModify = getNodeByIndex(index);
+        T returnValue = nodeToModify.item;
+        nodeToModify.item = value;
+        return returnValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T returnValue = unlink(getNodeByIndex(index));
+        size--;
+        return returnValue;
```

Comment:
якщо хочеш, то цей код можна спростити якщо пернести `size--;` в unlink() метод. Тоді код буде мати вигляд:
```suggestion
        return unlink(getNodeByIndex(index));
```

Reply:
done